TF influence function

katmeld opened the pull request:
fixes #119  
Added function to load the mavisp database into moonlight.
Added function to evaluate TF influence

mtiberti commented on NAMESPACE:
```diff
@@ -66,18 +71,25 @@ importFrom(graphics,text)
 importFrom(randomForest,randomForest)
 importFrom(readr,parse_guess)
+importFrom(readr,read_csv)
+importFrom(rlang,set_names)
```
this package should be added to `Imports:` in `DESCRIPTION`

mtiberti commented on NAMESPACE:
```diff
@@ -90,5 +102,6 @@ importFrom(tidyr,unnest)
 importFrom(utils,data)
 importFrom(utils,setTxtProgressBar)
 importFrom(utils,txtProgressBar)
+importFrom(withr,with_options)
```
same for this

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' 
+#' This function finds mutations in the transcription factors (TF) of the DEGs that have a TF in the database.
+#' 
+#' @param dataTRRUST RDA object of gene-TF pairs
```
`RDA` just refers to the file format R uses to store data; we can be a bit more specific, e.g. is this supposed to be a tibble?

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' \item TF (HUGO symbol of TF)
+#' \item GENE (HUGO symbol of gene)
+#' }
+#' @param dataMAF A MAF file RDA object
```
same for here

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' \item HGVSp_Short eg. p.V83F
+#' }
+#' @param dataDEGs Output DEA function
+#' @param dataMAVISp Output loadMAVISp function
```
these need to be a bit more descriptive

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' }
+#' @param dataDEGs Output DEA function
+#' @param dataMAVISp Output loadMAVISp function
+#' 
```
a description of `dataTFexpr` is missing

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' \itemize{
+#' \item TF (HUGO symbol of TF)
+#' \item GENE (HUGO symbol of gene)
+#' }
```
according to the code it should also contain a `InteractionType` col

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+    # dataMAF
+    if (is.null(dim(dataMAF))) {
+        stop("The mutation data must be a non-empty table")
+    }
```
`dim(dataMAF)` can be not-null in an empty dataframe:

```
> df <- data.frame(a=c())
> df
data frame with 0 columns and 0 rows
> dim(df)
[1] 0 0
> is.null(dim(df))
[1] FALSE
```

you can do something like:

```
if (is.null(df) | nrow(df) == 0) ...
```

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    # dataDEGs
+    if (is.null(dim(dataDEGs))) {
```
same comment as the other case

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    # dataTFexpr
+    if (!(dataTFexpr == FALSE) & is.null(dim(dataTFexpr))) {
```
I'm not sure why `dataTFexpr` defaults to `FALSE` if it is supposed to contain some data. Should it maybe default to `NULL`?

mtiberti replied:
Not sure if this is what you're going for here, but maybe it's a bit easier to:

```
if (dataTFexpre != FALSE & ! is.null(dim(dataTFexpr))) # if it's not false and not null, then ...
...
```

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+
+    # Load data --------------------------------
+    # Read maf and add ID number to each mutation
+    dataMAFFiltered <- dataMAF |> 
```
first time I see `|>` being used 😎

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+    return(mavisp_mut_DEG_TF)
+}
+
+mavispFiltering <- function(dataMAVISp){
+    # Keep only the stability classification
+    dataMAVISpFiltered <- dataMAVISp |>
+                    map(function(x) rename(x, 'stab_class' = matches('(Stability classification, [A-Za-z0-9]+, \\(Rosetta, FoldX\\))'))) |>
+                    keep(function(x) 'stab_class' %in% colnames(x)) |>
+                    map(function(x) select(x, 1, stab_class) |>
+                                    rename('mutation' = 1)) |>
+                    rbindlist(idcol = 'protein') |>
+                    as_tibble()
+    return(dataMAVISpFiltered)
+}
```
I don't see why this should should be a function since we use it only once, and it's not user-facing

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+mavispFiltering <- function(dataMAVISp){
+    # Keep only the stability classification
+    dataMAVISpFiltered <- dataMAVISp |>
+                    map(function(x) rename(x, 'stab_class' = matches('(Stability classification, [A-Za-z0-9]+, \\(Rosetta, FoldX\\))'))) |>
```
we should probably discuss if we want to support RaSP since we are gearing towards using it more

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -0,0 +1,76 @@
+#' 
+#' This function loads the MAVISp database from the directory specified by the user. 
+#' 
+#' @param mavispDB path to the MAVISp database
+#' @param proteins_of_interest vector containing specific proteins of interest in HUGO format
```
we could add an option to select mode, defaulting to simple mode

mtiberti commented on R/TFinfluence.R:
```diff
@@ -0,0 +1,147 @@
+#' \item tf_mutation
+#' \item stab_class (the effect on stability as classified by MAVISp)
+#' }
+#' 
```
this requires a working `@examples` section

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -0,0 +1,76 @@
+#' @export
+#' @examples
+#' 
+#' mavisp_data <- loadMAVISp(mavispDB = "/data/raw_data/computational_data/mavisp_database/biorxiv_v4_17112023",
+#'           proteins_of_interest = c('NQO1','TP53'))
+#' 
+#' mavisp_data <- loadMAVISp(mavispDB = "/data/raw_data/computational_data/mavisp_database/biorxiv_v4_17112023") 
```
we should consider on how to do these examples - examples should be runnable. We can either include a dataset with the package, or consider if we want to make the whole MAVISp dataset as a data package from BioC which would also be useful in general to the R community

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -0,0 +1,76 @@
+        rawFiles <- list.files(table_location,
+                               full.names = TRUE)
+
+        mavispData <- rawFiles |>
+                        set_names(str_split_i(basename(rawFiles), '-', 1)) |>
+                        # Supress non-fatal warnings
+                        map(function(x) withr::with_options(
+                                        list(rlib_name_repair_verbosity = "quiet"),
+                                        suppressWarnings(
+                                        classes = 'vroom_parse_issue',
+                                        read_csv(file = x,
+                                                progress = FALSE,
+                                                show_col_types = FALSE))))
```
this part is duplicated below; I think we can change code structure so that this is not necessary

mtiberti reviewed (changes requested):
- did we ever determine whether we are allowed to distribute TRRUST data with our package? If not we should just not include it
  -

katmeld commented:
I made a lot of changes and hopefully addressed all the small ones. Let me know if I missed some.

Some bigger changes:
- Included use of the PRA output (as intended) 
- Removed dataTFexpression. I think it would be better to approve the rest first, and include it later. I think we might need to discuss the data possibilities first.

I don’t think we settled on whether it could be included in the data or not. We can just remove it. The function does not load from the data directory.

There is no control for the MAVISp data. I am bit unsure how to tackle it. Ideas would be appreciated.

I have added something for the TFinfluence example, but I think we might need to discuss how to do that example and also the one for loadMAVISp

mtiberti commented on R/TFinfluence.R:
```diff
@@ -67,81 +104,71 @@ TFinfluence <- function(dataTRRUST,
+                    # Choose rosetta consensus, if it is there
+                    map(function(x) mutate(x, 'stab_class' = ifelse(test = !('stab_class_ros' %in% colnames(x)),
+                                                                 yes = stab_class_rasp,
+                                                                 no = stab_class_ros)) |>
```
here I think I would rather have the user decide which classification they want to use, using a function argument  - especially now that we are working towards going all-mutations for our MAVISp entries. This is because we often have a trade-off between the number of mutations we have available with Rosetta vs RaSP that is always a complete scan. So depending on what we are doing and what proteins we have we might prefer one or the other. I think we can default to Rosetta though as you do, because in general mutations in the MAF files are going to be subset of the ones we have in MAVISp as we discussed

mtiberti commented on R/TFinfluence.R:
```diff
@@ -67,81 +104,71 @@ TFinfluence <- function(dataTRRUST,
-    return(mavisp_mut_DEG_TF)
-}
+    filtered_effect <- drivers_mut_mavisp |>
+        filter((InteractionType == 'Activation' & stab_class == 'Destabilizing' & logFC_target < 0) |
+                (InteractionType == 'Repression' & stab_class == 'Destabilizing' & logFC_target > 0) |
```
In general - are our input DEGs supposed to be only significant DEGs (e.g. FDR < threshold & LogFC > threshold), as I understand? if so we should state it very clearly in the documentation, otherwise we risk the data pulling in and using data that is not relevant.

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
 #' This function loads the MAVISp database from the directory specified by the user. 
 #' 
-#' @param mavispDB path to the MAVISp database
+#' @param mavispDB path to the MAVISp database as a string
```
we can add a little bit more info of where the format comes from, since it's not at all standard - you can refer to our database OSF and the MAVISp preprint

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
+#' \item simple
+#' \item ensemble
+#'} 
+#' @param simulations string specifying the simulation to use. The simulations possible can be found in the MAVISp database.
```
this is not which simulation - it's the structural ensemble that was used to gather the MAVISp data. An ensemble can come from straight up a simulation or can be a collection of structures obtained in some other way.

also, it should be "the possible ensembles" or "the available ensembles" (swapped words)

mtiberti replied:
we should also mention that this is ignored in simple mode

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
 #' @export
 #' @examples
 #' 
 #' mavisp_data <- loadMAVISp(mavispDB = "/data/raw_data/computational_data/mavisp_database/biorxiv_v4_17112023",
```
we must have runnable examples because of BioC, and this won't work - it refers to a path in our own system. If you need data to run your examples or test on, it needs to be included in the R package. Of course it can be a tiny subset of the full database, since it's just an example.

Consider that every example is run automatically during the automated package verification and this fails if the examples are not runnable. They can be set as not runnable in some circumstances (e.g. the we cannot have csscape_somatic in the package), but in general they should be

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
-                       proteins_of_interest = NULL){
+                       proteins_of_interest = NULL,
+                       mode = 'simple',
+                       simulation = 'md'){
```
argument name to be changed (see previous comment)

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
+        }
+        table_location <- str_c(mavispDB,'/ensemble_mode/dataset_tables/')
+    } else {
+        stop('Mode not specified correctly. Takes values "simple" or "ensemble"')
```
s/Takes/Accepts/

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
-                                         filepath)) |>
-                            pull(filepath)
+                               as_tibble_col(column_name = 'filepath') |>
+                    filter(grepl(paste(proteins_of_interest, 
+                                        collapse = '|'),
+                                        filepath)) |>
```
do we risk grepping unrelated proteins with this? 

if e.g. we have `PRT12` in our `proteins_of_interest` and `PRT1` in our table

katmeld replied:
Nope, took care of that a few lines above by adding a hyphen. Had the issue with TP53 :)

mtiberti commented on R/loadMAVISp.R:
```diff
@@ -2,75 +2,95 @@
+                                show_col_types = FALSE))))
+    
+    # Combine stability results based on user specification
+    if (mode == 'ensemble' & length(simulation) == 1){
```
not sure of the second part of the if - we checked the same at the beginning of the function

mtiberti reviewed (changes requested):
- we need to add test, ideally at least one test per functions. These should check that the function is run and returns expected output. See the `tests` folder - we use the `testthat` package for this
- we also need to add a section in our vignette with the new functions, that showcase how to use them, and what we can get. Notice that the examples in the vignette must also be able to run automatically, in the same way the examples do - see `Moonlight2R.Rmd` in the `vignette` folder
- also notice that the whole build of the package must happen in a short time (<10 min) for us to pass the automated check 
- I've added you as author in the package
- can you add a short description of this new functionality in the `DESCRIPTION` file?
- I have run the BioC tests and we have some warning and errors that we need to fix. 
WARNINGS:
```
* checking Rd \usage sections ... WARNING
Undocumented arguments in Rd file 'loadMAVISp.Rd'
  ‘simulation’
Documented arguments not in \usage in Rd file 'loadMAVISp.Rd':
  ‘simulations’
```
ERRORS:
```
Running examples in ‘Moonlight2R-Ex.R’ failed
The error most likely occurred in:

> ### Name: TFinfluence
> ### Title: TFinfluence
> ### Aliases: TFinfluence
>
> ### ** Examples
>
>
> TFinfluence(dataTRRUST = dataTRRUST,
+            dataMAF = dataMAF,
+            dataDEGs = dataDEGs,
+            dataPRA = data_PRA,
+            dataMAVISp = mavisp_data)
Error: object 'data_PRA' not found
```
(the latter is connected with examples not being runnable)
